Dockerfile

```diff
@@ -64,10 +64,23 @@ RUN apk add --update --no-cache libpq tzdata yaml ${EXTRA_PACKAGES} && \
     cp /usr/share/zoneinfo/Europe/London /etc/localtime && \
     echo "Europe/London" > /etc/timezone
 
+# Create non-root user and group with specific UIDs/GIDs
+RUN addgroup -S appgroup -g 20001 && adduser -S appuser -G appgroup -u 10001
+
 # Copy files generated in the builder image
 COPY --from=builder /app /app
 COPY --from=builder /usr/local/bundle/ /usr/local/bundle/
 
+# Change ownership only for directories that need write access
+RUN chown -R appuser:appgroup /app/tmp /app/public/api/docs /app/log
+
+# Make everything read-only (with execute for directories)
+RUN chmod -R a-w /app && \
```

Comment:
We don't need to run any of the chmods. The default umask means the appuser has the right permissions. Plus we want to keep this as simple as possible.